src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs

```diff
@@ -1252,6 +1262,39 @@ private static DkmClrCompilationResultFlags GetLocalResultFlags(LocalSymbol loca
                 isIteratorOrAsyncMethod = GeneratedNames.GetKind(containingType.Name) == GeneratedNameKind.StateMachineType;
             }
 
+            var parameterNames = PooledHashSet<string>.GetInstance();
+            var parameterNamesInOrder = ArrayBuilder<string>.GetInstance();
+            // For version before .NET 4.5, we cannot find the sourceMethod properly:
+            // The source method coincides with the original method in the case.
+            // Here, we have to get parameters from the containingType.
+            // This does not guarantees the proper order of parameters.
+            if (isIteratorOrAsyncMethod && method == sourceMethod)
+            {
+                Debug.Assert(IsDisplayClassType(containingType));
+                foreach (var field in containingType.GetMembers().OfType<FieldSymbol>())
+                {
+                    // All iterator and async state machine fields (including hoisted locals) have mangled names, except
+                    // for hoisted parameters (whose field names are always the same as the original source parameters).
+                    var fieldName = field.Name;
+                    if (GeneratedNames.GetKind(fieldName) == GeneratedNameKind.None)
+                    {
+                        parameterNamesInOrder.Add(fieldName);
+                        parameterNames.Add(fieldName);
+                    }
+                }
+            }
+            else
+            {
+                foreach (var p in sourceMethod.Parameters)
+                {
+                    parameterNamesInOrder.Add(p.Name);
+                    parameterNames.Add(p.Name);
+                }
+            }
+
+            sourceMethodParametersInOrder = parameterNamesInOrder.ToImmutableArray();
+            parameterNamesInOrder.Free();
```

Comment:
Consider populating `parameterNames` outside of the `if`, from `parameterNamesInOrder`:
```
foreach (var name in parameterNames)
{
    parameterNames.Add(name);
}
```

Reply:
Sounds good. Thanks!

---
In reply to: [135916177](https://github.com/dotnet/roslyn/pull/21707#discussion_r135916177) [](ancestors = 135916177)